// dbd.js/package/functions/funcs/createRole.js
const keyPerms = require("../../utils/permissions")

module.exports = async d => {
    const code = d.command.code

    const r = code.split("$createRole[").length - 1

    const inside = code.split("$createRole[")[r].split("]")[0]

    const fields = inside.split(";")

    const name = fields.shift()

    const color = fields.shift()

    const mentionable =fields.shift()

    const hoisted = fields.shift()

    const position = fields.shift()

    const permissions = Object.entries(keyPerms).filter(perm => fields.includes(perm[0])).map(perm => perm[1])
    
    const role = await d.message.guild.roles.create({
        data: {
            name: name,
            color: color,
            mentionable: mentionable === "yes",
            hoist: hoisted === "yes",
            position: position,
            permissions: permissions
        }
    }).catch(err => {})

    if (!role) return d.error(`:x: Failed to create role!`)

    return {
        code: code.replaceLast(`$createRole[${inside}]`, "")
    }
}
