// dbd.js/package/functions/funcs/newTicket.js
const embed = require("../../handlers/errors")

module.exports = async d => {
    
    const code = d.command.code
    
    const r = code.split("$newTicket").length - 1
    
    const inside = code.split("$newTicket[")[r].split("]")[0]
    
    const [
        name, 
        msg, 
        categoryID, 
        returnID = "no", 
        error
    ] = inside.split(";")
    
    const channel = await d.message.guild.channels.create(name, {
        parent: categoryID || undefined 
    }).catch(err => null)
    
    if (!channel) return embed(d, error) 
    
    await d.client.db.set({
        varID: channel.id,
        variable: "tickets", 
        value: "open"
    })
    
    await channel.updateOverwrite(d.message.author.id, {
        SEND_MESSAGES: true, 
        VIEW_CHANNEL: true, 
        READ_MESSAGE_HISTORY: true 
    }).catch(err => null)
    
    if (msg) {
        const copy = {...d}
        
        copy.message.channel = channel 
        
        embed(copy, msg) 
    }
    
    
    return {
        code: code.replaceLast(`$newTicket[${inside}]`, returnID === "yes" ? channel.id : "")
    }
}
