// dbd.js/package/utils/memberOptions.js
module.exports = {
    id: "The ID of the user;.id", 
    name: "The name of this user;.user.usernane.deleteBrackets()", 
    guildID: "The ID of the guild this member was updated on;.guild.id",
    nick: "The nickname of this user before it was updated, if any;.nicknane || ''", 
    roles: "The role names of the user before it was updated;.roles.filter(r => r.name !== '@everyone').map(r => r.name.deleteBrackets()).join(', ')", 
    partial: "Whether the member structure is cached or not;.partial", 
    premiumStamp: "The timestamp this user started boosting the server, or 0 if they're not boosting;.premiumSinceTimestamp || 0", 
    joinedStamp: "The timestamp when this user joined the server;.joinedTimestamp",
    voiceID: "The ID of the voice channel this user's in;.voice.channelID || ''", 
    displayHex: "Returns the hex color of this user's highest role;.displayHexColor", 
    highestRoleID: "The ID of the highest role of this user in the guild;.roles.highest.id",
    permissions: "The permissions for this member;.permissions.toArray().goof('_')", 
    bannable: "Whether if the user is bannable by the client or not;.bannable",
    kickable: "Whether if the user is kickable by the client or not;.kickable",
    manageable: "Whether if the user can be managed by the client or not;.manageable", 
    displayColor: "Displays the color of the highest role in this user;.displayColor",
    status: "The status for this user;.presence.status",
    activities: "The activities for this user;.activities.map(c => c.name).join(', ')", 
    
}
